File: kernel/drivers/hrand.c
#include "drivers/base.h"
#include "drivers/hrand.h"
#include "util/stdlib.h"

/*
Based on this source from the linux kernel
https://github.com/thenaran/linux-rpi/blob/master/drivers/char/hw_random/bcm2708-rng.c
And this
https://github.com/bztsrc/raspi3-tutorial/blob/master/06_random/rand.c
Haven't been able to find any actual documentation for this.
*/

#define RNG_BASE (PBASE + 0x00104000)
#define RNG_CTRL (RNG_BASE)
#define RNG_STATUS (RNG_BASE + 0x4)
#define RNG_DATA (RNG_BASE + 0x8)

// For enabling rgen
// The 2nd options is faster but less random
#define RNG_RBGEN 0x1
#define RNG_RBG2X 0x2

// Initial numbers are less random
#define RNG_WARMUP_COUNT 0x40000

void hrand_init(void) {
  // NOTE: Other srcs mask interrupts here?
  put32(RNG_STATUS, RNG_WARMUP_COUNT);
  put32(RNG_CTRL, RNG_RBGEN);
}

/*
  Returns a large random number.
  Unclear what RAND_MAX is.
  NOTE: There is no checking that hrand is actually enabled.
  QUESTION: What prevents two rapid reads from getting the same number here?
            Is the underlying peripheral aware of the read?
*/
uint32_t hrand(void) {
  // Wait for some entropy
  while (!(get32(RNG_STATUS) >> 24))
    ;
  return get32(RNG_DATA);
}

File: kernel/src/drivers/hrand.rs
//! Module to interface with hardware randomness
//!
//! Based on these sources [1], [2]
//!
//! [1]: https://github.com/thenaran/linux-rpi/blob/master/drivers/char/hw_random/bcm2708-rng.c
//! [2]: https://github.com/bztsrc/raspi3-tutorial/blob/master/06_random/rand.c

use drivers;
use rand_core;
use spin::Mutex;

const BASE: usize = drivers::P_BASE + 0x0010_4000;
const CTRL: *mut u32 = BASE as *mut u32;
const STATUS: *mut u32 = (BASE + 0x4) as *mut u32;
const DATA: *const u32 = (BASE + 0x8) as *const u32;

// Used when enabling rand
const NORMAL_INIT: u32 = 0x1;
// This option is faster but less random
const FAST_INIT: u32 = 0x2;

// Initial numbers are less random
const WARMUP_COUNT: u32 = 0x4_0000;

lazy_static! {
    /// Global rand generator
    pub static ref RAND: Mutex<HRand> = Mutex::new(HRand::init(false));
}

pub struct HRand;

impl HRand {
    fn init(fast: bool) -> HRand {
        // @NOTE: Other srcs mask interrupts here?
        unsafe {
            if fast {
                STATUS.write_volatile(WARMUP_COUNT);
                CTRL.write_volatile(FAST_INIT);
            } else {
                // Faster, less random
                STATUS.write_volatile(WARMUP_COUNT);
                CTRL.write_volatile(NORMAL_INIT);
            }
        }
        HRand
    }
}

impl rand_core::RngCore for HRand {
    /// Returns a random u32.
    /// This call will spin
    /// @TODO: Check that hrand is enabled
    /// @QUESTION: What prevents two rapid reads from getting the same number here?
    fn next_u32(&mut self) -> u32 {
        unsafe {
            while (STATUS.read_volatile() >> 24) == 0 {}
            DATA.read_volatile()
        }
    }

    fn next_u64(&mut self) -> u64 {
        rand_core::impls::next_u64_via_fill(self)
    }

    fn fill_bytes(&mut self, dest: &mut [u8]) {
        rand_core::impls::fill_bytes_via_next(self, dest)
    }

    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), rand_core::Error> {
        self.fill_bytes(dest);
        Ok(())
    }
}
